zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -14,161 +21,294 @@ use crate::service::finalized_state::{
 
 /// Runs disk format upgrade for adding Sapling and Orchard note commitment subtrees to database.
 ///
+/// Trees are added to the database in reverse height order, so that wallets can sync correctly
+/// while the upgrade is running.
+///
 /// Returns `Ok` if the upgrade completed, and `Err` if it was cancelled.
 #[allow(clippy::unwrap_in_result)]
+#[instrument(skip(upgrade_db, cancel_receiver))]
 pub fn run(
     initial_tip_height: Height,
     upgrade_db: &ZebraDb,
     cancel_receiver: &mpsc::Receiver<CancelFormatChange>,
 ) -> Result<(), CancelFormatChange> {
-    let mut subtree_count = 0;
-    let mut prev_tree: Option<_> = None;
-    for (height, tree) in upgrade_db.sapling_tree_by_height_range(..=initial_tip_height) {
-        // Return early if there is a cancel signal.
+    // # Consensus
+    //
+    // Zebra stores exactly one note commitment tree for every block with sapling notes.
+    // (It also stores the empty note commitment tree for the genesis block, but we skip that.)
+    //
+    // The consensus rules limit blocks to less than 2^16 sapling and 2^16 orchard outputs. So a
+    // block can't complete multiple level 16 subtrees (or complete an entire subtree by itself).
+    // Currently, with 2MB blocks and v4/v5 sapling and orchard output sizes, the subtree index can
+    // increase by 1 every ~20 full blocks.
```

Comment:
###### Optional Suggestion

_"full blocks_" can also refer to blocks with no notes.

```suggestion
    // increase by at most 1 every ~20 blocks.
```